service.go

```diff
@@ -133,9 +129,32 @@ func (opts *Opts) applyDefaults(ctx context.Context) error {
 	rng := mrand.New(mrand.NewSource(srand.MustSecure())) // nolint:gosec // we need to use math/rand here, but it is seeded from crypto/rand
 
 	opts.applyDefaultsGetDatastore()
+	opts.applyPushDefaults()
+
+	if opts.RootDatastore == nil && opts.DatastoreDir != "" {
+		bopts := badger.DefaultOptions
+		bopts.ValueLogLoadingMode = options.FileIO
 
-	if err := opts.applyPushDefaults(); err != nil {
-		return err
+		ds, err := badger.NewDatastore(opts.DatastoreDir, &bopts)
+		if err != nil {
+			return fmt.Errorf("unable to init badger datastore: %w", err)
+		}
+		opts.RootDatastore = ds
+
+		oldClose := opts.close
+		opts.close = func() error {
+			var err error
+			if oldClose != nil {
+				err = oldClose()
+			}
+
+			if dserr := ds.Close(); err == nil && dserr != nil {
+				// only return ds error if no error has been caught earlier
+				err = fmt.Errorf("unable to close datastore: %w", dserr)
+			}
```

Comment:
check for `errs = multierr.Append(errs, err)`
You can grep the project to find some examples.

Reply:
Done. Let me know if I should do it in a different way.